sql/sql_optimizer.cc

```diff
@@ -2472,10 +2474,27 @@ static bool can_switch_from_ref_to_range(THD *thd, JOIN_TAB *tab,
 
     calc_length_and_keyparts(tab->position()->key, tab,
                              tab->position()->key->key, tab->prefix_tables(),
-                             NULL, &length, &keyparts, &dep_map, &maybe_null);
+                             nullptr, &length, &keyparts, &dep_map, &maybe_null);
     if (!maybe_null &&  // 3)
         !dep_map)       // 4)
     {
+      /*
+        If there was a quick select that
+         - was scanning the same range
+         - used MRR
+         - MRR implementation indicated that it's preferred to use MRR over
+           ref(const).
+        then switch to range access as it will read rows using the MRR interface
+      */
+      if (tab->position()->key->key == tab->quick()->index &&
+          length == tab->quick()->max_used_key_length &&
+          tab->quick()->get_type() == QUICK_SELECT_I::QS_TYPE_RANGE) {
+        auto *qr = static_cast<QUICK_RANGE_SELECT*>(tab->quick());
+        if (!(qr->mrr_flags & HA_MRR_USE_DEFAULT_IMPL) &&
+            (qr->mrr_flags & HA_MRR_CONVERT_REF_TO_RANGE))
```

Comment:
Avoid implicit conversions
```suggestion
            (qr->mrr_flags & HA_MRR_CONVERT_REF_TO_RANGE) != 0)
```